diff --git a/src/main.cpp b/src/main.cpp
index 9ec5002..cbea83a 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -78,6 +78,7 @@ private:
         createLogicalDevice();
         createSwapChain();
         createImageViews();
+        createGraphicsPipeline();
     }
 
     void mainLoop() {
@@ -424,6 +425,13 @@ private:
         }
     }
     /////////////////////////////////////////////////////////////////
+
+    /////////////////////////////////////////////////////////////////
+    /// pipeline
+    void createGraphicsPipeline() {
+
+    }
+    /////////////////////////////////////////////////////////////////
 };
 
 int main() {
