diff --git a/src/main.cpp b/src/main.cpp
index 5638a66..392cdef 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -65,6 +65,9 @@ private:
     std::vector<vk::raii::Framebuffer> m_swapChainFramebuffers;
     vk::raii::CommandPool m_commandPool{ nullptr };
     std::vector<vk::raii::CommandBuffer> m_commandBuffers;
+    vk::raii::Semaphore m_imageAvailableSemaphore{ nullptr };
+    vk::raii::Semaphore m_renderFinishedSemaphore{ nullptr };
+    vk::raii::Fence m_inFlightFence{ nullptr };
     /////////////////////////////////////////////////////////////////
     
     /////////////////////////////////////////////////////////////////
@@ -95,12 +98,16 @@ private:
         createFramebuffers();
         createCommandPool();
         createCommandBuffer();
+        createSyncObjects();
     }
 
     void mainLoop() {
         while (!glfwWindowShouldClose( m_window )) {
             glfwPollEvents();
+            drawFrame();
         }
+        
+        m_device.waitIdle();
     }
 
     void cleanup() {
@@ -479,6 +486,19 @@ private:
         vk::RenderPassCreateInfo renderPassInfo;
         renderPassInfo.setAttachments( colorAttachment );
         renderPassInfo.setSubpasses( subpass );
+
+        vk::SubpassDependency dependency;
+        dependency.srcSubpass = VK_SUBPASS_EXTERNAL;
+        dependency.dstSubpass = 0;
+
+        dependency.srcStageMask = vk::PipelineStageFlagBits::eColorAttachmentOutput;
+        dependency.srcAccessMask = {};
+
+        dependency.dstStageMask = vk::PipelineStageFlagBits::eColorAttachmentOutput;
+        dependency.dstAccessMask = vk::AccessFlagBits::eColorAttachmentWrite;
+
+        renderPassInfo.setDependencies( dependency );
+
         
         m_renderPass = m_device.createRenderPass(renderPassInfo);
     }
@@ -702,6 +722,49 @@ private:
         commandBuffer.end();
     }
     /////////////////////////////////////////////////////////////////
+
+    /////////////////////////////////////////////////////////////////
+    /// render and present
+    void createSyncObjects() {
+        vk::SemaphoreCreateInfo semaphoreInfo;
+        vk::FenceCreateInfo fenceInfo(
+            vk::FenceCreateFlagBits::eSignaled  // flags
+        );
+
+        m_imageAvailableSemaphore = m_device.createSemaphore( semaphoreInfo );
+        m_renderFinishedSemaphore = m_device.createSemaphore( semaphoreInfo );
+        m_inFlightFence = m_device.createFence( fenceInfo );
+    }
+    void drawFrame() {
+        m_device.waitForFences( *m_inFlightFence, true, UINT64_MAX );
+        m_device.resetFences( *m_inFlightFence );
+
+        uint32_t imageIndex = m_swapChain.acquireNextImage(UINT64_MAX, m_imageAvailableSemaphore).second;
+
+        m_commandBuffers[0].reset();
+        recordCommandBuffer(m_commandBuffers[0], imageIndex);
+
+        vk::SubmitInfo submitInfo;
+
+        submitInfo.setWaitSemaphores( *m_imageAvailableSemaphore );
+        std::vector<vk::PipelineStageFlags> waitStages { vk::PipelineStageFlagBits::eColorAttachmentOutput };
+        submitInfo.setWaitDstStageMask( waitStages );
+
+        submitInfo.setCommandBuffers( *m_commandBuffers[0] );
+
+        submitInfo.setSignalSemaphores( *m_renderFinishedSemaphore );
+
+        m_graphicsQueue.submit(submitInfo, m_inFlightFence);
+
+        vk::PresentInfoKHR presentInfo;
+        presentInfo.setWaitSemaphores( *m_renderFinishedSemaphore );
+
+        presentInfo.setSwapchains( *m_swapChain );
+        presentInfo.pImageIndices = &imageIndex;
+
+        m_presentQueue.presentKHR( presentInfo );
+    }
+    /////////////////////////////////////////////////////////////////
 };
 
 int main() {
